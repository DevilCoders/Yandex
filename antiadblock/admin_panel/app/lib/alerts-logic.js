import PriorityQueue from 'js-priority-queue';

const COOEFF_STATE_PRIO = 10000000000;
const WEIGHTS = {
        red: 1,
        yellow: 2,
        green: 3
    };

function statePriority(state) {
    return WEIGHTS[state] || Number.MAX_SAFE_INTEGER;
}

function alertPriorityComparator(a, b) {
    return (((statePriority(a.state) - statePriority(b.state)) * COOEFF_STATE_PRIO) + (b.transition_time - a.transition_time));
}

function alertFilter(item) {
    return (statePriority(item.state) < WEIGHTS.green);
}

export function prepareServiceAlerts(alerts, limit = 5) {
    const queue = new PriorityQueue({comparator: alertPriorityComparator});

    let result = {
        topAlerts: [],
        tailAlerts: []
    };

    for (const alertItem of alerts) {
        if (alertFilter(alertItem)) {
            queue.queue(alertItem);
        }
    }

    const len = queue.length;
    for (let i = 0; i < len; i++) {
        if (i < limit) {
            result.topAlerts.push(queue.dequeue());
        } else {
            result.tailAlerts.push(queue.dequeue());
        }
    }

    return result;
}

export function prepareAlerts(alertItems, limit = 5) {
    // do prepare for income alerts
    let topAlerts = {},
        tailAlerts = {};

    for (const [serviceId, serviceAlertItems] of Object.entries(alertItems)) {
        const serviceAlerts = prepareServiceAlerts(serviceAlertItems, limit);

        topAlerts[serviceId] = serviceAlerts.topAlerts;
        tailAlerts[serviceId] = serviceAlerts.tailAlerts;
    }
    return {topAlerts, tailAlerts};
}
// todo: add tests for this
