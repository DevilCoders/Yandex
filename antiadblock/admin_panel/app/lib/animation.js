const easeOutQuadSimple = t => {
    return t * (2 - t);
};

const easeOutQuad = (currentTime, fromValue, toValue, fullTime) => {
    const simpleTime = currentTime / fullTime;
    const coeff = easeOutQuadSimple(simpleTime);

    return fromValue + ((toValue - fromValue) * coeff);
};

const SCROLL_OFFSET = 30;
const SCROLL_PERIOD = 10;

let scrollFunction,
    animationMap = new Map();

scrollFunction = (element, offset, delay, cb) => {
    const fromValue = element.scrollTop;
    const toValue = offset - SCROLL_OFFSET;

    let currentTime = 0,
        interval = animationMap.get(element),
        stopAnimation = () => {
            clearInterval(interval);
            animationMap.delete(element);
            element.removeEventListener('wheel', stopAnimation);
            if (typeof cb === 'function') {
                cb();
            }
        };

    if (interval) {
        stopAnimation();
    }

    interval = setInterval(() => {
        currentTime += SCROLL_PERIOD;
        element.scrollTop = easeOutQuad(currentTime, fromValue, toValue, delay);
        if (currentTime === delay) {
            stopAnimation();
        }
    }, SCROLL_PERIOD);
    animationMap.set(element, interval);

    element.addEventListener('wheel', stopAnimation);
};

export const scroll = scrollFunction;
