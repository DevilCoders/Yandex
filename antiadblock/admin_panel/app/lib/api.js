import ResponseError from 'app/errors/response-error';

const getRequestConfiguration = () => {
    return {
        queryParams: {}
    };
};

export default function request(prefix, requestObject) {
    const reqQueryParams = requestObject.queryParams || {};
    const prepQueryParams = Object
        .keys(reqQueryParams)
        .reduce((queryParams, key) => {
            if (reqQueryParams[key] !== undefined) {
                queryParams[key] = reqQueryParams[key];
            }
            return queryParams;
    }, {});
    const queryParams = Object.assign({},
        prepQueryParams,
        getRequestConfiguration().queryParams
    );
    let {options, path} = requestObject,
        params;

    options = options || {};

    // Можно передать массив в queryParams, тогда он должен склеиться в param=1&param=2&param=3 итд
    params = Object.keys(queryParams).map(function(paramName) {
        const paramValue = queryParams[paramName];

        if (Array.isArray(paramValue)) {
            return queryParams[paramName].map(paramItemValue => `${paramName}=${paramItemValue}`).join('&');
        }

        return `${paramName}=${encodeURIComponent(paramValue)}`;
    }).join('&');

    options.headers = options.headers || {};

    if (options.body && !(options.body instanceof FormData)) {
        options.headers['Content-Type'] = 'application/json';
        options.body = JSON.stringify(options.body);
    }

    return fetch(prefix + path + '?' + params, Object.assign({}, {
        credentials: 'same-origin'
    }, options)).then(function(response) {
        if (response.ok) {
            if (options.json !== false) {
                return response
                    .json()
                    .catch(function() {
                        return Promise.reject(new ResponseError(response.status, response));
                    });
            }

            return response.text();
        }

        return response.json()
            .then(function(json) {
                return Promise.reject(new ResponseError(response.status, json));
            }, function() {
                return Promise.reject(new ResponseError(response.status));
            });
    }, function(error) {
        return Promise.reject(new ResponseError(null, error));
    });
}

export function antiadbRequest(requestObject, countRetry = 0) {
    return request('/rest', requestObject).then(data => {
        return data;
    }, err => {
        return countRetry > 0 ?
            antiadbRequest(requestObject, countRetry--) :
            Promise.reject(err);
    });
}
