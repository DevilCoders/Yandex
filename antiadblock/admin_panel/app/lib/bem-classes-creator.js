const ELEM_PREFIX = '__';
const MOD_PREFIX = '_';
const MOD_VALUE_PREFIX = '_';

function createMods(block, mods) {
    let classNames = [];

    if (mods) {
        Object.keys(mods).forEach(mod => {
            if (mods[mod]) {
                const modString = mods[mod] === true ?
                    mod :
                    `${mod}${MOD_VALUE_PREFIX}${mods[mod]}`;

                classNames.push(`${block}${MOD_PREFIX}${modString}`);
            }
        });
    }

    return classNames;
}

function bemClasses(block, elem, mods) {
    const blockElem = `${block}${elem ? `${ELEM_PREFIX}${elem}` : ''}`;

    return [blockElem, ...createMods(blockElem, mods)];
}

export default function bemClassesCreator(block, elem, mods, mix) {
    let classNames = bemClasses(block, elem, mods);

    if (mix) {
        if (Array.isArray(mix)) {
            mix.forEach(mixItem => {
                if (mixItem) {
                    classNames = classNames.concat(bemClasses(mixItem.block, mixItem.elem, mixItem.mods));
                }
            });
        } else {
            classNames = classNames.concat(bemClasses(mix.block, mix.elem, mix.mods));
        }
    }

    return classNames.join(' ');
}
