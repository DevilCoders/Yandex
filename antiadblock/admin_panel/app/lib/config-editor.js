import ConfigTextEditor from 'app/components/config-editor/__text/config-editor__text';
import ConfigTupleEditor from 'app/components/config-editor/__tuple/config-editor__tuple';
import ConfigEnumEditor from 'app/components/config-editor/__enum/config-editor__enum';
import ConfigCheckboxGroupEditor from 'app/components/config-editor/__checkbox-group/config-editor__checkbox-group';
import ConfigArrayEditor from 'app/components/config-editor/__array/config-editor__array';
import ConfigTokensEditor from 'app/components/config-editor/__tokens/config-editor__tokens';
import ConfigTagsEditor from 'app/components/config-editor/__tags/config-editor__tags';
import ConfigNumberEditor from 'app/components/config-editor/__number/config-editor__number';
import ConfigObjectEditor from 'app/components/config-editor/__object/config-editor__object';
import ConfigBoolEditor from 'app/components/config-editor/__bool/config-editor__bool';
import ConfigLinkEditor from 'app/components/config-editor/__link/config-editor__link';
import ConfigCryptEditor from 'app/components/config-editor/__crypt/config-editor__crypt';
import ConfigElemsEditor from 'app/components/config-editor/__elems/config-editor__elems';
import ConfigJsonEditor from 'app/components/config-editor/__json/config-editor__json';
import ConfigHtmlEditor from 'app/components/config-editor/__html/config-editor__html';

import {get, isEqual, set} from 'lodash';
import {STATUS} from 'app/enums/config';

const typeMapping = {
    select: ConfigEnumEditor,
    multi_select: ConfigEnumEditor,
    multi_checkbox: ConfigCheckboxGroupEditor,
    regexp: ConfigTextEditor,
    array: ConfigArrayEditor,
    tokens: ConfigTokensEditor,
    tags: ConfigTagsEditor,
    string: ConfigTextEditor,
    number: ConfigNumberEditor,
    object: ConfigObjectEditor,
    bool: ConfigBoolEditor,
    link: ConfigLinkEditor,
    crypt_body: ConfigCryptEditor,
    detect_elems: ConfigElemsEditor,
    detect_custom: ConfigJsonEditor,
    detect_html: ConfigHtmlEditor,
    test_data: ConfigJsonEditor,
    json: ConfigJsonEditor
};

export function getConfig(config) {
    const answer = {
        constructor: typeMapping[config.type] || ConfigTextEditor,
        props: {
            placeholder: config.placeholder,
            hint: config.hint,
            title: config.title
        }
    };

    let child;

    // custom config
    switch (config.type) {
        // TODO костыль пока не будет выпилен совсем ANTIADB-2130
        case 'test_data':
            answer.props.title = 'test-data-title';
            answer.props.hint = 'test-data-hint';
            answer.props.placeholder = 'test-data-placeholder';
            break;
        case 'select':
            answer.props.enum = config.values;
            break;
        case 'multi_checkbox':
        case 'multi_select':
            answer.props.enum = config.values;
            answer.props.multi = true;
            break;
        case 'object':
            answer.props.Item = ConfigTupleEditor;
            answer.props.childProps = {
                items: [typeMapping[config.key], typeMapping[config.value]],
                placeholder: [
                    config.placeholder.key_placeholder,
                    config.placeholder.value_placeholder
                ]
            };
            break;
        case 'tokens':
        case 'array':
            child = getConfig(config.children);
            answer.props.childProps = child.props;
            answer.props.Item = child.constructor;
            break;
        case 'tags':
            answer.props.placeholder = config.children.placeholder;
            break;
        default:
    }

    return answer;
}

export function getDefaultValue(schema, name) {
    for (let i = 0; i < schema.length; i++) {
        const items = schema[i].items;
        for (let j = 0; j < items.length; j++) {
            const item = items[j].key;

            if (item.name === name) {
                return item.default;
            }
        }
    }
    return undefined;
}

export function isEmptyValue(type, value, schema, name) {
    switch (type) {
        case 'select':
            return getDefaultValue(schema, name) === value;
        case 'array':
        case 'tags':
        case 'multi_checkbox':
        case 'tokens':
            return !value.length;
        case 'object':
        case 'json':
        case 'test_data':
            return !value && !Object.keys(value).length;
        case 'bool':
            return !value;
        case 'detect_html':
        case 'string':
            return !value;
        case 'number':
            return isNaN(value);
        default:
            return true;
    }
}

export function getSchemaTypeAndDefVal(schema) {
    return schema.reduce((obj, group) => {
        group.items.forEach(item => {
            obj.values[item.key.name] = item.key.default;
            obj.types[item.key.name] = item.key.type_schema.type;
        });

        return obj;
    }, {
        values: {},
        types: {}
    });
}

export function mergeConfig(parent, current, dataSettings, schema) {
    const mergedConfig = {};

    for (let [key, valueP] of Object.entries(parent)) {
        if (get(dataSettings, [key, 'UNSET'], false)) {
            mergedConfig[key] = schema[key];
        } else {
            mergedConfig[key] = (current[key] !== undefined && current[key] !== null) ? current[key] : valueP;
        }
    }

    return mergedConfig;
}

// TODO убрать эту функцию, когда иеархические конфиги приведут в порядок
// На текущий момент с бека могут прилетать в текущем конфиге данные,
// которые равняются дефолтным значениям схемы, это последствия старых конфигов.
// После создания иеархии проблема должна уйти
// Предположительно, можно убрать функцию через месяц (8 марта 2020)
export function clearDefaultValues(dataCurrent = {}, dataSettings = {}, schema = {}) {
    return Object.keys(dataCurrent).reduce((newData, key) => {
        if (!isEqual(dataCurrent[key], schema[key]) && !get(dataSettings, [key, 'UNSET'], false)) {
            set(newData, key, get(dataCurrent, key));
        }

        return newData;
    }, {});
}

export function getStatus(statuses, isExp) {
    return isExp ? STATUS.EXPERIMENT : statuses.reduce((status, item) => (
        item === STATUS.ACTIVE ? item :
            item === STATUS.TEST && status !== STATUS.ACTIVE ? item : status
    ), '');
}
