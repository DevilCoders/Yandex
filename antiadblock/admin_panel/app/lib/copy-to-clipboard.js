export default function (value) {
    const control = document.createElement('textarea');

    control.style = 'position: absolute; left: -1000px; top: -1000px';
    control.value = value || '';
    document.body.appendChild(control);
    control.select();
    document.execCommand('copy');
    document.body.removeChild(control);
}
