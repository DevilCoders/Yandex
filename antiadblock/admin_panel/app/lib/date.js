import i18n from 'app/lib/i18n';
import dayjs from 'dayjs';
import 'dayjs/locale/ru';
import weekOfYear from 'dayjs/plugin/weekOfYear';

dayjs.locale('ru');
dayjs.extend(weekOfYear);

export function getMonthNamesForLocale(locale) {
    const format = new Intl.DateTimeFormat(locale, {month: 'short'});
    const months = [];
    for (let month = 0; month < 12; month++) {
        const testDate = new Date(Date.UTC(2000, month, 1, 0, 0, 0));
        months.push(format.format(testDate));
    }
    return months;
}

export function formatDateTo(dateString, format) {
    return dayjs(dateString).format(format);
}

export function timeAgo(date) {
    const now = new Date();
    const human = ['second', 'minute', 'hour', 'day', 'week', 'month', 'year'];
    const delta = [1000, 60, 60, 24, 7, 4, 12];

    let diffrence = Math.abs(date.getTime() - now.getTime());

    for (let i = 0; i < delta.length; i++) {
        diffrence /= delta[i];

        // less than
        if (diffrence < 1) {
            return `${i18n('datetime', 'less-than')} ${i18n('datetime', human[i], 2)} ${i18n('datetime', 'ago')}`;
        }

        if (i === delta.length - 1 || delta[i + 1] > diffrence) {
            const int = Math.round(diffrence);
            return `${int} ${i18n('datetime', human[i], int)} ${i18n('datetime', 'ago')}`;
        }
    }
}

export {
    dayjs
};

