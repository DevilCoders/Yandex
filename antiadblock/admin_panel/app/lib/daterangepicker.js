export function isBetween(minDate, maxDate, selectedDay) {
    return !isNotBetween(minDate, maxDate, selectedDay);
}

export function isBetweenExclusive(startDate, endDate, day) {
    return startDate && endDate && day > startDate && day < endDate;
}

export function isNotBetween(minDate, maxDate, selectedDay) {
    return (minDate !== null && selectedDay.isBefore(minDate, 'day')) || (maxDate !== null && selectedDay.isAfter(maxDate, 'day'));
}

export function isWeekend(day) {
    return day.day() === 0 || day.day() === 6;
}

export function isoWeek(d) {
    d = new Date(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()));
    d.setUTCDate(d.getUTCDate() + 4 - (d.getUTCDay() || 7));
    const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
    const msInDay = 86400000;

    return Math.ceil((((d - yearStart) / msInDay) + 1) / 7);
}

export function getValueOnInterVal(left, right, value) {
    let res = value;

    if (value < left) {
        res = left;
    } else if (right < value) {
        res = right;
    }

    return res;
}

export function dateDayJsIsEqual(date1, date2) {
    if (!date1 || !date2) {
        return false;
    }

    return (
        date1.date() === date2.date() &&
        date1.month() === date2.month() &&
        date1.year() === date2.year()
    );
}
