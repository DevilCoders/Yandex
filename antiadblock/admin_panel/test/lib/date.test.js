import {timeAgo, getMonthNamesForLocale, formatDateTo} from 'app/lib/date';

// mock env and translation module
process.env.BEM_LANG = 'ru';
jest.mock('app/i18n/datetime/datetime.i18n/ru', () => ({
    datetime: {
        ago: 'назад',
        'less-than': 'меньше',
        day: ['день', 'дня', 'дней', 'дней'],
        month: ['месяц', 'месяца', 'месяцев', 'месяцев'],
        hour: ['час', 'часа', 'часов', 'часов'],
        second: ['секунду', 'секунды', 'секунд', 'секунд'],
        year: ['год', 'года', 'лет', 'лет']
    }
}));

describe('date', () => {
    describe('timeAgo', () => {
        test('should return less than a second ago', () => {
            const now = new Date();
            expect(timeAgo(now)).toMatchSnapshot();
        });

        test('should return 5 months ago', () => {
            const now = new Date();
            now.setMonth(now.getMonth() - 5);
            expect(timeAgo(now)).toMatchSnapshot();
        });

        test('should return 5 years ago', () => {
            const now = new Date();
            now.setFullYear(now.getFullYear() - 5);
            expect(timeAgo(now)).toMatchSnapshot();
        });

        test('should return 1 day ago', () => {
            const now = new Date();
            now.setHours(now.getHours() - 35);
            expect(timeAgo(now)).toMatchSnapshot();
        });
    });

    describe('formatDateTo', () => {
        test('set invalid date will return \'Invalid Date\'', () => {
            const time = 'asdadsa';
            const format = 'DD.MM.YYYY HH:mm:ss';
            const result = 'Invalid Date';

            expect(formatDateTo(time, format)).toEqual(result);
        });

        test('set date will return in format \'DD.MM.YYYY HH:mm:ss\'', () => {
            const time = '2019-09-11T15:22:54';
            const format = 'DD.MM.YYYY HH:mm:ss';
            const result = '11.09.2019 15:22:54';

            expect(formatDateTo(time, format)).toEqual(result);
        });
    });

    describe('getMonthNamesForLocale', () => {
        test('test en locale', () => {
            expect(getMonthNamesForLocale('en')).toEqual(['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']);
        });
    });
});
