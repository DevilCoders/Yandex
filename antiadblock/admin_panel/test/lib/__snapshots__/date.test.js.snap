// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`date timeAgo should return 1 day ago 1`] = `"1 день назад"`;

exports[`date timeAgo should return 5 months ago 1`] = `"5 месяцев назад"`;

exports[`date timeAgo should return 5 years ago 1`] = `"5 лет назад"`;

exports[`date timeAgo should return less than a second ago 1`] = `"меньше секунды назад"`;
